### PR TITLE
EKS Addon Fargate Bug Fix

### PR DESCRIPTION
**Pull request by Paramadon**
*Issue #, if available:*
This pr fixes this issue: https://github.com/aws/amazon-cloudwatch-agent-operator/issues/183

The pr https://github.com/aws-observability/helm-charts/pull/41 added the default tolerations for all pod workloads installed by the EKS add-on, the CWA daemonset and fluent-bit daemonset now tolerate all taints ensuring they run on every single node on the cluster.

For clusters setup with a combination of managed node groups and fargate profiles, this causes pods to be brought up even on the fargate nodes even though there exists a taint on these nodes - which causes the daemonsets to never complete deployment and the add-on goes into a failed state.

This pr fixes this using node affinity and making sure that the any pod with key eks.amazonaws.com/compute-type and with the value: fargate, is not scheduled.

Why can't daemonset work on fargate nodes?

DaemonSets do not work on Fargate nodes because Fargate nodes are serverless. DaemonSets are designed to run a copy of a pod on every node in a Kubernetes cluster, but Fargate nodes do not have a fixed or consistent set of nodes like traditional EC2-based Kubernetes clusters. Since you cannot control or ensure the placement of pods on specific Fargate nodes, the fundamental requirement of DaemonSets to run on every node cannot be met.

*Description of changes:*

Tested changes by running:
`helm template amazon-cloudwatch-observability ./amazon-cloudwatch-observability  --include-crds --namespace amazon-cloudwatch --set clusterName=cloudwatchagent-operator-toleration-fargate --set region=us-east-1 | kubectl apply --namespace amazon-cloudwatch -f -`

Just adding this to the daemonsets to avoid getting scheduled on fargate nodes:

```
        affinity:
          nodeAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
              nodeSelectorTerms:
              - matchExpressions:
                - key: eks.amazonaws.com/compute-type
                  operator: NotIn
                  values:
                  - fargate
```

Fargate nodes have this taint: amazonaws.com/compute-type=fargate:NoSchedule, so the above affinity will block off any daemonset pods that are being scheduled on fargate nodes.

### Before bug fix - pods are pending which results in eks addon failure

Screenshot 2024-07-01 at 4 27 59 PM

### After bug fix - all pods are running and eks addon status is active

Screenshot 2024-07-02 at 10 54 14 AM

### Picture of what was added to Cloudwatch-agent, dcgm exporter, fluent bit, and neuron monitor daemonset
Screenshot 2024-07-02 at 12 14 34 PM (2)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by musa-asad — DISMISSED** [review]

File: charts/amazon-cloudwatch-observability/templates/linux/neuron-monitor-daemonset.yaml [hunk 1]
```diff
@@ -29,38 +33,36 @@ spec:
       cpu: 256m
       memory: 128Mi
   env:
-  - name: NODE_NAME
-    valueFrom:
-      fieldRef:
-        fieldPath: spec.nodeName
-  - name: PATH
-    value: /usr/local/bin:/usr/bin:/bin:/opt/aws/neuron/bin
-  - name: GOMEMLIMIT
```
Comment (movence) [inline charts/amazon-cloudwatch-observability/templates/linux/neuron-monitor-daemonset.yaml:38]: why is this being removed?

File: charts/amazon-cloudwatch-observability/templates/linux/cloudwatch-agent-daemonset.yaml [hunk 2]
```diff
@@ -43,74 +52,74 @@ spec:
...
-    readOnly: true
-  - mountPath: /dev/disk
-    name: devdisk
-    readOnly: true
-  - mountPath: /etc/amazon-cloudwatch-observability-agent-cert
-    name: agenttls
-    readOnly: true
-  - mountPath: /var/lib/kubelet/pod-resources
-    name: kubelet-podresources
+    - mountPath: /rootfs
```
Comment (movence) [inline charts/amazon-cloudwatch-observability/templates/linux/cloudwatch-agent-daemonset.yaml:55]: can we revert this indentation changes?

**Review by musa-asad — APPROVED** [review]

File: charts/amazon-cloudwatch-observability/values.yaml [hunk 3]
```diff
@@ -565,7 +567,7 @@ dcgmExporter:
       us-gov-east-1: 743662458514.dkr.ecr.us-gov-east-1.amazonaws.com
       us-gov-west-1: 743662458514.dkr.ecr.us-gov-west-1.amazonaws.com
   configmap: dcgm-exporter-config-map
-  arguments: ["--web-config-file=/etc/dcgm-exporter/web-config.yaml"]
+  arguments: ["--web-config-file=/etc/dcgm-exporter/web-config.yaml" ]
```
Comment (movence) [inline charts/amazon-cloudwatch-observability/values.yaml:570]: I think auto indent is not perfect for sure

**Review by movence — APPROVED** [review]

**Comment by Abdullahtheauthor** [comment]
Are you sure that this fix is deployed correctly to the helm chart? I tried to install it using the instructions in the readme file, the helm isn't updated with this fix.

**Comment by sky333999** [comment]
> Are you sure that this fix was deployed correctly to the helm chart? I tried to install it using the instructions in the readme file, the helm isn't updated with this fix. @Paramadon

Hi @Abdullahtheauthor .. this change is yet to be released. So if you followed the steps in the readme, you likely installed the latest released version of the helm chart that wouldnt have this change.

**Comment by Abdullahtheauthor** [comment]
That's exactly what happened. I had to clone the repo locally to be able to use the fix.
Thanks for the update.

**Pull request merged**